src/app/CodeRepo/lib/backends/github/index.js

```diff
@@ -0,0 +1,149 @@
+"use strict";
+
+const log = require("log");
+const rp = require("request-promise");
+const moment = require("moment");
+const { AsyncEventEmitter } = require("emitter");
+const GithubEventEmitter = require("./github_event_emitter");
+
+const GITHUB_API_BASE = "https://api.github.com";
+const GITHUB_BASE = "https://github.com";
+
+/*
+To get started with github:
+1. blalbabla
+2. blalalalala
+*/
+
+class GithubBackend extends AsyncEventEmitter {
+    constructor(id, backend, Repository, Revision) {
+        super();
+        this.id = id;
+        this.backend = backend;
+        this.locks = {};
+        this.Repository = Repository;
+        this.Revision = Revision;
+
+        this.githubEmitter = new GithubEventEmitter();
+    }
+
+    async start() {
+        // Start event monitoring towards github
+
+        try {
+            await this._startMonitorEventStream();
+        } catch (err) {
+            log.error("Failed to setup local web server: ", err);
+        }
+    }
+
+    async validateRepository(/* event, data */) {
+        // TODO: Validate gerrit specific options
+    }
+
+    async _onPing(event) {
+        console.log("ping event received");
+    }
+
+    async _onPullRequestOpen(event) {
+        console.log("pull-request-open received");
+        console.log(event);
+
+        const changeId = event.pull_request.head.sha;
+        const repositoryId = event.repository.name;
+        const repository = await this.Repository.findOne({ _id: repositoryId });
+        if (repository) {
+            const ref = {
+                index: 1,
+                email: null,
+                name: event.pull_request.user.login,
+                submitted: moment.unix(event.pull_request.created_at).utc().format(),
+                comment: event.pull_request.title,
+                change: {
+                    oldrev: null,
+                    newrev: null,
+                    refname: null // Use event.refName all the time instead?
```

Comment:
oldrev, newrev and refname is important.
newrev is the specific commit SHA.
refname is the refname used when pulling the remote changes.